crypto/ec/ecp_sm2p256.c

```diff
@@ -0,0 +1,816 @@
+#include "ec_local.h"
+
+#if defined(__GNUC__)
+# define ALIGN32        __attribute((aligned(32)))
+# define ALIGN64        __attribute((aligned(64)))
+#elif defined(_MSC_VER)
+# define ALIGN32        __declspec(align(32))
+# define ALIGN64        __declspec(align(64))
+#else
+# define ALIGN32
+# define ALIGN64
+#endif
+
+#define P256_LIMBS      (256/BN_BITS2)
+
+extern const BN_ULONG ecp_sm2p256_precomputed[8 * 32 * 256];
+
+typedef struct {
+    BN_ULONG X[P256_LIMBS];
+    BN_ULONG Y[P256_LIMBS];
+    BN_ULONG Z[P256_LIMBS];
+} P256_POINT;
+
+typedef struct {
+    BN_ULONG X[P256_LIMBS];
+    BN_ULONG Y[P256_LIMBS];
+} P256_POINT_AFFINE;
+
+/* Coordinates of G, for which we have precomputed tables */
+static const BN_ULONG def_xG[P256_LIMBS] = {
+    0x715a4589334c74c7, 0x8fe30bbff2660be1,
+    0x5f9904466a39c994, 0x32c4ae2c1f198119
+};
+
+static const BN_ULONG def_yG[P256_LIMBS] = {
+    0x02df32e52139f0a0, 0xd0a9877cc62a4740,
+    0x59bdcee36b692153, 0xbc3736a2f4f6779c,
+};
+
+/* p and order for SM2 according to GB/T 32918.5-2017 */
+static const BN_ULONG def_p[P256_LIMBS] = {
+    0xffffffffffffffff, 0xffffffff00000000,
+    0xffffffffffffffff, 0xfffffffeffffffff
+};
+static const BN_ULONG def_ord[P256_LIMBS] = {
+    0x53bbf40939d54123, 0x7203df6b21c6052b,
+    0xffffffffffffffff, 0xfffffffeffffffff
+};
+
+/* p_div_2 = (p + 1) / 2 */
+static const BN_ULONG p_div_2[P256_LIMBS] ALIGN32 = {
+    0x8000000000000000, 0xffffffff80000000,
+    0xffffffffffffffff, 0x7fffffff7fffffff
+};
+
+/* ord_div_2 = (ord + 1) / 2 */
+static const BN_ULONG ord_div_2[P256_LIMBS] ALIGN32 = {
+    0xa9ddfa049ceaa092, 0xb901efb590e30295,
+    0xffffffffffffffff, 0x7fffffff7fffffff
+};
+
+static const BN_ULONG ONE[P256_LIMBS] = {1, 0, 0, 0};
+
+/* Functions implemented in assembly */
+/*
+ * Most of below mentioned functions *preserve* the property of inputs
+ * being fully reduced, i.e. being in [0, modulus) range. Simply put if
+ * inputs are fully reduced, then output is too.
+ */
+/* Right shift: a >> 1 */
+void bn_rshift1(BN_ULONG *a);
+/* Modular div by 2: r = a/2 mod p */
+void ecp_sm2p256_div_by_2(BN_ULONG *r, const BN_ULONG *a);
+/* Modular add: r = a+b mod p */
+void ecp_sm2p256_add(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular add: r = a+b mod n, where n = ord(p) */
+void ecp_sm2p256_add_mod_ord(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular sub: r = a-b mod p */
+void ecp_sm2p256_sub(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular sub: r = a-b mod n, where n = ord(p) */
+void ecp_sm2p256_sub_mod_ord(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular mul by 3: out = 3*a mod p */
+void ecp_sm2p256_mul_by_3(BN_ULONG *r, const BN_ULONG *a);
+/* Modular mul: r = a*b mod p */
+void ecp_sm2p256_mul(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular sqr: r = a^2 mod p */
+void ecp_sm2p256_sqr(BN_ULONG *r, const BN_ULONG *a);
+
+static ossl_inline BN_ULONG is_zero(BN_ULONG a)
```

Comment:
These would be better added to internal constant time header I think.

Reply:
I replaced `is_zero()` with `constant_time_is_zero_64()` which is defined in constant_time.h, is it OK?